### ISSUE TITLE
Compile error with glslang version 8 on Linux

### ISSUE THREAD

**Issue by rhabacker** [vsg-dev/VulkanSceneGraph#407]
**Describe the bug**
Compiling vsg with the mentioned vulkan header version fails with the following errors:
```
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp: In member function 'bool vsg::ShaderCompiler::compile(vsg::ShaderStages&, const std::vector<std::__cxx11::basic_string<char> >&, vsg::ref_ptr<const vsg::Options>)':
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:160:24: error: 'EShLangRayGen' was not declared in this scope
[  319s]              envStage = EShLangRayGen;
[  319s]                         ^~~~~~~~~~~~~
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:160:24: note: suggested alternative: 'EShLangRayGenNV'
[  319s]              envStage = EShLangRayGen;
[  319s]                         ^~~~~~~~~~~~~
[  319s]                         EShLangRayGenNV
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:164:24: error: 'EShLangAnyHit' was not declared in this scope
[  319s]              envStage = EShLangAnyHit;
[  319s]                         ^~~~~~~~~~~~~
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:164:24: note: suggested alternative: 'EShLangAnyHitNV'
[  319s]              envStage = EShLangAnyHit;
[  319s]                         ^~~~~~~~~~~~~
[  319s]                         EShLangAnyHitNV
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:168:24: error: 'EShLangClosestHit' was not declared in this scope
[  319s]              envStage = EShLangClosestHit;
[  319s]                         ^~~~~~~~~~~~~~~~~
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:168:24: note: suggested alternative: 'EShLangClosestHitNV'
[  319s]              envStage = EShLangClosestHit;
[  319s]                         ^~~~~~~~~~~~~~~~~
[  319s]                         EShLangClosestHitNV
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:172:24: error: 'EShLangMiss' was not declared in this scope
[  319s]              envStage = EShLangMiss;
[  319s]                         ^~~~~~~~~~~
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:172:24: note: suggested alternative: 'EShLangMissNV'
[  319s]              envStage = EShLangMiss;
[  319s]                         ^~~~~~~~~~~
[  319s]                         EShLangMissNV
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:176:24: error: 'EShLangIntersect' was not declared in this scope
[  319s]              envStage = EShLangIntersect;
[  319s]                         ^~~~~~~~~~~~~~~~
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:176:24: note: suggested alternative: 'EShLangIntersectNV'
[  319s]              envStage = EShLangIntersect;
[  319s]                         ^~~~~~~~~~~~~~~~
[  319s]                         EShLangIntersectNV
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:180:24: error: 'EShLangCallable' was not declared in this scope
[  319s]              envStage = EShLangCallable;
[  319s]                         ^~~~~~~~~~~~~~~
[  319s] /home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/src/vsg/vk/ShaderCompiler.cpp:180:24: note: suggested alternative: 'EShLangCallableNV'
[  319s]              envStage = EShLangCallable;
[  319s]                         ^~~~~~~~~~~~~~~
[  319s]                         EShLangCallableNV
[  319s] make[2]: *** [src/vsg/CMakeFiles/vsg.dir/build.make:2111: src/vsg/CMakeFiles/vsg.dir/vk/ShaderCompiler.cpp.o] Error 1
[  319s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/VulkanSceneGraph-0.2.8+20220222+git.c220f4ca/build'
[  319s] make[2]: *** Waiting for unfinished jobs....
```

**To Reproduce**
1. Inspect the appended build log for SLE_15_SP3, which has been extracted from https://build.opensuse.org/package/show/home:rhabacker:branches:games/libvsg
2. See error

**Expected behavior**
libvsg should compile with that vulkan version  

**Desktop:**
 - OS: SLE_15_SP3

[_log.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/8132545/_log.txt)

**Comment by robertosfield** [comment]
Feck.  Looks like they just moved the goal posts deliberately breaking the
API.  I can't go chasing this right now, but once I've got the Allocator
work done I'll try out 1.3 and see if I can find a convenient way through
the mess they are creating.

**Comment by sbrkopac** [comment]
I took some time today and cleaned up how we are handling CI #408. I also added the 1.3.204.0 SDK to ubuntu but it doesn't seem to throw the same error? I also updated to the latest vulkan version locally and it seems fine on windows.

**Comment by rhabacker** [comment]
>  I also added the 1.3.204.0 SDK to ubuntu but it doesn't seem to throw the same error? 

Did you compile with glslang enabled ? The related code is only included, when compiling with glslang development package installed, see https://github.com/vsg-dev/VulkanSceneGraph/blob/c220f4caba410c8d97eb8a98173b47cebe6b5b85/src/vsg/vk/ShaderCompiler.cpp#L100

**Comment by sbrkopac** [comment]
I believe so: 
```
-- The following OPTIONAL packages have been found:

 * vsg_glslang
```
Unless I need to enable it via a cmake flag? I just checked on my local windows machine and I can see that code path being compiled.

**Comment by sbrkopac** [comment]
Ah, I see. The ```#HAS_KHR_RAYTRACING``` isn't enabled so I'm not hitting the same code path you are. I will have to review in the morning unless you want to take a look at my PR.

**Comment by rhabacker** [comment]
>  * vsg_glslang

No, if building with glslang is enabled, in the cmake log there need to be
```
[   79s] -- Found glslang: /usr/lib64/libglslang.so
```

**Comment by rhabacker** [comment]
> vsg_glslang

This message does not indicate that glslang was found because it is part of the vsg source and is always found.

**Comment by rhabacker** [comment]
> I took some time today and cleaned up how we are handling CI #408. I also added the 1.3.204.0 SDK to ubuntu but it doesn't seem to throw the same error?

I checked the logs from the mentioned build and I can confirm that glslang has been enabled. The build failure is related to glslang, not to the vulkan version:
```
$ xxx@host:/home/osc/build-root/SLE_15_SP3-x86_64/usr/include/glslang> grep -rn EShLangRayGen
Public/ShaderLang.h:95:    EShLangRayGenNV,
Public/ShaderLang.h:113:    EShLangRayGenNVMask       = (1 << EShLangRayGenNV),
```
The failed build uses `glslang-devel-8.13.3559+git20-1.27`, while on the CI it uses `version: 11.6.0`. Completed builds for other distribution versions use `glslang-devel-11.8.0`. It follows that a minimal glslang version is required for building vsg.

**Comment by robertosfield** [comment]
Hi Ralf,

So from what I understand so far, the version of SUSE you are using has an
older version of glslang already installed, and this is being found by the
VSG rather than the modern one provided by the VulkanSDK?

Are you using the VulkanSDK?  Are you setting the VULAN_SDK env var?

Cheers,
Robert.Message ID: ***@***.***
com>
>
>

**Comment by rhabacker** [comment]
Building vsg on distributions, as is done for osg for example, depends on packages that provide needed sdk's. The SDK provided by Lunar is a monolithic entity consisting of vulkan-header, vulkan-loader, glslang, spir, shaderc and possibly other parts. In distributions, these individual parts are divided into corresponding packages.In openSUSE, respectively, in order to build vsg, I need to install the following packages: vulkan-devel, glslang-devel, spir,-tools and shaderc. Since distributions provide additional package sources from which packages can be installed, or the base distribution does not provide a newer version, it can happen that older versions of packages are installed, which happened here with the distribution SLE_15_SP3 (SUSE Linux Enterprise 15 Service Pack3) and vsg did not check whether the glslang version used is sufficient.

**Comment by rhabacker** [comment]
BTW: glslang since 11.0.0 provides a way to get the actual version from an installed header (see https://github.com/KhronosGroup/glslang/commit/fbe9a23baf2cb020fe3f80d68d972e2a97d9b954)

**Comment by robertosfield** [comment]
On Fri, 25 Feb 2022 at 10:30, Ralf Habacker ***@***.***>
wrote:

> Building vsg on distributions, as is done for osg for example, depends on
> packages that provide needed sdk's. The SDK provided by Lunar is a
> monolithic entity consisting of vulkan-header, vulkan-loader, glslang,
> spir, shaderc and possibly other parts. In distributions, these individual
> parts are divided into corresponding packages.In openSUSE, respectively, in
> order to build vsg, I need to install the following packages: vulkan-devel,
> glslang-devel, spir,-tools and shaderc. Since distributions provide
> additional package sources from which packages can be installed, or the
> base distribution does not provide a newer version, it can happen that
> older versions of packages are installed, which happened here with the
> distribution SLE_15_SP3 (SUSE Linux Enterprise 15 Service Pack3) and vsg
> did not check whether the glslang version used is sufficient.Message ID:
> ***@***.***>
>
GLSLang is a bit of a nightmare set of disparate versions that do things in
different ways making it really hard to juggle which version provides what,
even the reporting of the API version is all over the place and breaks
across versions.

The VSG has it's own vsg_glslangConfig.cmake to try and juggle the mess and
keep things compiling.

Your usage case is throwing in another curve ball.  I have enough other
tasks to handle right now to not get tangled in the GLSLang/SUSE mess.

**Comment by vsg-dev** [comment]
I have finally got on to improving the VSG's compatibility with older Vulkan versions (now compiles against Vulkan 1.1.70 headers found in Ununtu 18.04) and have begun working on getting the VSG to compile against older versions of glslang.  This is work can be found in the [ImprovedBackwardsCompatibility](https://github.com/vsg-dev/VulkanSceneGraph/compare/ImprovedBackwardsCompatibility) branch.  The commit with fixes for older glslag version is:
17217ae1075ae01f8cfdbba68259e03ca774203e

I haven't got the VSG examples to compile with older glslang yet as it looks like the lib dependencies changes from version to version. I don't have access to SUSE but have found Ubuntu 20.04's glslang debian package is an old version of glang and appears to have the same compile issues as reported above.  The compile errors have now been resolved, so it's just the linking issues left to resolve.  

This may help fix the SUSE issues, but it may require further work, on that front I'd appreciate testing of the fixes.

**Comment by robertosfield** [comment]
@rhabacker could you try out VSG master to see if it resolves the compile errors on SUSE?

**Comment by rhabacker** [comment]
Hi Robert,
I am on vacation right now. I can take a look next week.

**Comment by rhabacker** [comment]
I started a build from git master branch commit 7aa0f39 using `libglslang-suse9-8.13.3559+git20-lp152.1.5` and got the following configuring configure error:
```
[   65s] -- Failed to find glslang
```
but building did not fail as shown in the log file [_log.txt](https://github.com/vsg-dev/VulkanSceneGraph/files/9001768/_log.txt)

**Comment by rhabacker** [comment]
> it won't be able to compile GLSL shaders to SPIR-V

maybe it would be good to inform users about this limitation e.g. 

```
-- Failed to find glslang  - support to compile GLSL shaders to SPIR-V is not available
```

> but otherwise will be fine.

Than this bug `Compile error with glslang version 8 on Linux` seems to be fixed.

**Comment by robertosfield** [comment]
Thanks for the testing.  I've added an extra message and compilation/shader composition not being available when glslang is not available: 56f6432d79d9ac18ea0f775d878aeb887ebaf9bc

**Comment by rhabacker** [comment]
nice - works as expected :-)
